scaladoc/src/dotty/tools/scaladoc/renderers/Renderer.scala

```diff
@@ -45,7 +45,52 @@ abstract class Renderer(rootPackage: Member, val members: Map[DRI, Member], prot
           val msg = s"ERROR: Multiple index pages for doc found ${indexes.map(_.file)}"
           report.error(msg)
 
-        val templatePages = templates.map(templateToPage(_, siteContext))
+        // Below code is for walking in order the tree and modifing its nodes basing on its neighbours
+
+        // First we flatten templates to get them sorted in-order
+        def flattenedTemplates(template: LoadedTemplate): Seq[LoadedTemplate] =
+          template +: template.children.flatMap(flattenedTemplates)
+
+        // We add dummy guards
+        val allTemplates: Seq[Option[LoadedTemplate]] = None +: templates.flatMap(flattenedTemplates).map(Some(_)) :+ None
+
+        // Let's gather the list of maps for each template with its in-order neighbours
+        val newSettings: Seq[Map[String, Object]] = allTemplates.sliding(size = 3, step = 1).map {
+          case None :: None :: Nil =>
+            Map.empty
+          case prev :: mid :: next :: Nil =>
+            def link(sibling: Option[LoadedTemplate]): Option[String] =
+              def realPath(path: Path) = if Files.isDirectory(path) then Paths.get(path.toString, "index.html") else path
+              sibling.map { n =>
+                val realMidPath = realPath(mid.get.file.toPath)
+                val realSiblingPath = realPath(n.file.toPath)
+                realMidPath.relativize(realSiblingPath).toString.stripPrefix("../")
+              }
+            List(link(prev).map("previous" -> _), link(next).map("next" -> _)).flatten.toMap
+        }.toSeq
+
+        // We update the immutable tree of templates by walking in-order
+        def updateSettings(template: LoadedTemplate, additionalSettings: Seq[Map[String, Object]]): (LoadedTemplate, Seq[Map[String, Object]]) =
+          val head :: tail = additionalSettings
```

Comment:
Not super important, but it’s better to require a `List` if we do head/tail decomposition:

```suggestion
        def updateSettings(template: LoadedTemplate, additionalSettings: List[Map[String, Object]]): (LoadedTemplate, List[Map[String, Object]]) =
          val head :: tail = additionalSettings
```

Reply:
Ok,  will change